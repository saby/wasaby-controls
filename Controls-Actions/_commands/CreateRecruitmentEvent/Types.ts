/**
 * Информация о текущих настройках для создания события
 * @private
 */
export interface IEventInfo {
    /**
     * Запланировано ли событие, влияет на фильтр при запросе к источнику данных
     */
    isPlaned: boolean;
    /**
     * Тип события, берется с БЛ
     */
    type: string;
}
