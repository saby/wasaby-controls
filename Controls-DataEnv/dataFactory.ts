/**
 * Библиотека, содержащая фабрики данных.
 * @library
 * @public
 * @includes IDataConfig Controls-DataEnv/_dataFactory/interface/IDataConfig
 * @includes IDataFactory Controls-DataEnv/_dataFactory/interface/IDataFactory
 * @includes IDataFactory Controls-DataEnv/_dataFactory/interface/IBaseDataFactoryArguments
 */
export { IDataConfig, TDataConfigs } from './_dataFactory/interface/IDataConfig';
export { IDataFactory } from './_dataFactory/interface/IDataFactory';
export { IBaseDataFactoryArguments } from './_dataFactory/interface/IBaseDataFactoryArguments';

export {
    IAreaDataFactoryResult,
    IAreaDataFactory,
    IAreaDataFactoryArguments,
} from './_dataFactory/Area/IAreaDataFactory';
export { default as AreaSlice, IAreaState } from './_dataFactory/Area/Slice';
export { default as Area } from './_dataFactory/Area/Factory';
export { default as CompatibleArea } from './_dataFactory/Area/CompatibleFactory';
export { default as Custom } from './_dataFactory/Custom/Factory';
export { default as CompatibleCustom } from './_dataFactory/Custom/CompatibleFactory';
export { default as Form } from './_dataFactory/Form/Factory';
export { default as FormSlice, IFormSlice } from './_dataFactory/Form/Slice';
export { FormSliceActionType } from './_dataFactory/Form/IFormDataFactory';
export {
    ICustomDataFactoryArguments,
    ICustomDataFactory,
} from './_dataFactory/Custom/ICustomDataFactory';
