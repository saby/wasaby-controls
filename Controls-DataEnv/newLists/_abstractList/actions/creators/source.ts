/**
 * @kaizen_zone 039c82f1-a0a3-4548-82d6-c9e1dbaf5de0
 */
import type { source } from '../types';

/**
 * Конструктор действия, для загрузки предыдущей пачки данных.
 * @function
 * @return source.TLoadPrevAction
 */
export const loadPrev = (): source.TLoadPrevAction => ({
    type: 'prev',
    payload: {},
});

/**
 * Конструктор действия, для загрузки следующей пачки данных.
 * @function
 * @return source.TLoadNextAction
 */
export const loadNext = (): source.TLoadNextAction => ({
    type: 'next',
    payload: {},
});
