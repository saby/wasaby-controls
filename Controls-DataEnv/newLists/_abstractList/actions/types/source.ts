/**
 * @kaizen_zone 039c82f1-a0a3-4548-82d6-c9e1dbaf5de0
 */
import type { TAbstractAction } from 'Controls-DataEnv/dispatcher';

/**
 * Тип действия, для загрузки предыдущей пачки данных.
 */
export type TLoadPrevAction = TAbstractAction<'prev', {}>;

/**
 * Тип действия, для загрузки следующей пачки данных.
 */
export type TLoadNextAction = TAbstractAction<'next', {}>;

/**
 * Тип действий для работы с источником данных, доступные в любом списке,
 * независимо от типа ViewModel, к которой он подключен (web/mobile).
 */
export type TAnySourceAction = TLoadPrevAction | TLoadNextAction;
