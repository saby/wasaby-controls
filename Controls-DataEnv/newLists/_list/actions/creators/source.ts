/**
 * @kaizen_zone 039c82f1-a0a3-4548-82d6-c9e1dbaf5de0
 */
import type { IListSavedState } from 'Controls/dataSource';
import type { IBaseSourceConfig, INavigationSourceConfig } from 'Controls/interface';
import type { IListState } from '../../interface/IListState';
import type { CrudEntityKey } from 'Types/source';
import type { IReloadItemOptions } from 'Controls/listCommands';
import type { source } from '../types';

// FIXME: Типы действий описаны только чтобы собралась дока,
//  т.к. сами действия не те, что должны быть.
/**
 * Конструктор действия setSavedSourceState
 */
export const setSavedSourceState = (
    id: string,
    state: Partial<IListSavedState>
): source.TSetSavedSourceStateAction => ({
    type: 'setSavedSourceState',
    payload: {
        id,
        state,
    },
});

/**
 * Конструктор действия updateSavedState
 */
export const updateSavedState = (): source.TUpdateSavedSourceStateAction => ({
    type: 'updateSavedSourceState',
    payload: {},
});

/**
 * Конструктор действия reload
 */
export const reload = (
    sourceConfig?: INavigationSourceConfig,
    keepNavigation?: boolean,
    onResolve?: Function,
    onReject?: Function
): source.TReloadAction => ({
    type: 'reload',
    payload: {
        sourceConfig,
        keepNavigation,
        onResolve,
        onReject,
    },
});

/**
 * Конструктор действия newItemsReceived
 */
export const newItemsReceived = (
    payload: source.TNewItemsReceivedAction['payload']
): source.TNewItemsReceivedAction => ({
    type: 'newItemsReceived',
    payload,
});

/**
 * Конструктор действия complexUpdateSource
 */
export const complexUpdateSource = (
    prevState: IListState,
    nextState: IListState
): source.TComplexUpdateSourceAction => ({
    type: 'complexUpdateSource',
    payload: {
        prevState,
        nextState,
    },
});

/**
 * Конструктор действия load
 */
export const load = (sourceConfig?: IBaseSourceConfig): source.TLoadAction => ({
    type: 'load',
    payload: { sourceConfig },
});

/**
 * Конструктор действия oldSliceLoad
 */
export const oldSliceLoad = (
    payload: source.TOldSliceLoadAction['payload']
): source.TOldSliceLoadAction => ({
    type: 'oldSliceLoad',
    payload,
});

/**
 * Конструктор действия dataLoadedSuccess
 */
export const dataLoadedSuccess = (
    payload: source.TDataLoadedSuccessAction['payload']
): source.TDataLoadedSuccessAction => ({
    type: 'dataLoadedSuccess',
    payload,
});

/**
 * Конструктор действия fetch
 */
export const fetch = (): source.TFetchAction => ({
    type: 'fetch',
    payload: {},
});

/**
 * Конструктор действия requestFetch
 */
export const requestFetch = (): source.TRequestFetchAction => ({
    type: 'requestFetch',
    payload: {},
});

/**
 * Конструктор действия initSource
 */
export const initSource = (): source.TInitSourceAction => ({
    type: 'initSource',
    payload: {},
});

/**
 * Конструктор действия awaitAllRequests
 */
export const awaitAllRequests = (): source.TAwaitAllRequests => ({
    type: 'awaitAllRequests',
    payload: {},
});

/**
 * Конструктор действия loadOnSourceController
 */
export const loadOnSourceController = (
    payload: source.TLoadOnSourceControllerAction['payload']
): source.TLoadOnSourceControllerAction => ({
    type: 'loadOnSourceController',
    payload,
});

/**
 * Конструктор действия reloadOnSourceController
 */
export const reloadOnSourceController = (
    payload: source.TReloadOnSourceControllerAction['payload']
): source.TReloadOnSourceControllerAction => ({
    type: 'reloadOnSourceController',
    payload: {
        ...payload,
        isFirstLoad: !!payload.isFirstLoad,
    },
});

/**
 * Конструктор действия loadNodes
 */
export const loadNodes = (
    payload: source.TLoadNodesAction['payload']
): source.TLoadNodesAction => ({
    type: 'loadNodes',
    payload,
});

/**
 * Конструктор действия loadToDirectionOld
 */
export const loadToDirectionOld = (
    payload: source.TLoadToDirectionOldAction['payload']
): source.TLoadToDirectionOldAction => ({
    type: 'loadToDirectionOld',
    payload,
});

/**
 * Конструктор действия loadToDirectionNew
 */
export const loadToDirectionNew = (
    payload: source.TLoadToDirectionNewAction['payload']
): source.TLoadToDirectionNewAction => ({
    type: 'loadToDirectionNew',
    payload,
});

/**
 * Конструктор действия setPreloadedItems
 */
export const setPreloadedItems = (
    payload: source.TSetPreloadedItemsAction['payload']
): source.TSetPreloadedItemsAction => ({
    type: 'setPreloadedItems',
    payload,
});

/**
 * Конструктор действия resolveStateAfterUpdateItems
 */
export const resolveStateAfterUpdateItems = (
    payload: source.TResolveStateAfterUpdateItemsAction['payload']
): source.TResolveStateAfterUpdateItemsAction => ({
    type: 'resolveStateAfterUpdateItems',
    payload,
});

/**
 * Конструктор действия reloadItem
 */
export const reloadItem = (
    key: CrudEntityKey,
    options?: IReloadItemOptions,
    onResolve?: Function,
    onReject?: Function
): source.TReloadItemAction => ({
    type: 'reloadItem',
    payload: {
        key,
        options,
        onResolve,
        onReject,
    },
});

/**
 * Конструктор действия reloadItems
 */
export const reloadItems = (
    keys: CrudEntityKey[],
    onResolve?: Function,
    onReject?: Function
): source.TReloadItemsAction => ({
    type: 'reloadItems',
    payload: {
        keys,
        onResolve,
        onReject,
    },
});
