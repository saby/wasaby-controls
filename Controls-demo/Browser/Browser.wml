<div class="controlsDemo__wrapper">
    <div class="controlsDemo__cell">
        <Router.router:Reference attr:class="controls-text-label"
                                 state="/Controls-demo/app/:app"
                                 app="{{ 'Controls-demo/Browser/WithOperationPanel/Index' }}">
            <a href="{{content.href}}">WithOperationPanel</a>
        </Router.router:Reference>
        <Controls-demo.Browser.WithOperationPanel.Index/>
    </div>
</div>
