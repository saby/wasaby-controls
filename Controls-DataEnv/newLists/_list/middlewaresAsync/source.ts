/**
 * @kaizen_zone 039c82f1-a0a3-4548-82d6-c9e1dbaf5de0
 */
import { asyncMiddlewareFactory } from 'Controls-DataEnv/dispatcher';
import { TListMiddleware } from '../types/TListMiddleware';

export const sourceMiddleware: TListMiddleware = asyncMiddlewareFactory(
    'Controls/listWebReducers:source',
    'source',
    [
        'updateSavedSourceState',
        'setSavedSourceState',
        'reload',
        'load',
        'loadOnSourceController',
        'reloadOnSourceController',
        'newItemsReceived',
        'resolveStateAfterUpdateItems',
        'loadNodes',
        'oldSliceLoad',
        'reloadItem',
        'reloadItems',
        'loadToDirectionOld',
        'loadToDirectionNew',
        'setPreloadedItems',
        'dataLoadedSuccess',
        'requestFetch',
        'fetch',
        'initSource',
        'awaitAllRequests',
        'complexUpdateSource',
    ]
);
