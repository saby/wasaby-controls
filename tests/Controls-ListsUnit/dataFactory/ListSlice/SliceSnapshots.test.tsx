/**
 * @jest-environment jsdom
 */
import { unmountComponentAtNode } from 'react-dom';
import { IListState, ListSlice } from 'Controls/dataFactory';
import {
    createContext,
    useContext,
    useRef,
    useState,
    useMemo,
    createRef,
    forwardRef,
    useImperativeHandle,
    useCallback,
    useLayoutEffect,
} from 'react';
import { Memory } from 'Types/source';
import { RecordSet } from 'Types/collection';
import { act } from 'react-dom/test-utils';
import { render, waitFor } from '@testing-library/react';
import { WasabyEvents } from 'UICore/Events';
import {
    INavigationOptionValue,
    INavigationPageSourceConfig,
} from 'Controls/_interface/INavigation';
import { isEqual, clone } from 'Types/object';
import { Record } from 'Types/entity';
import { fetch, HTTPStatus } from 'Browser/Transport';

const flatData = [
    {
        key: 0,
        name: 'Sasha',
        department: 'Platform',
    },
    {
        key: 1,
        name: 'Sergey',
        department: 'Platform',
    },
    {
        key: 2,
        name: 'Maksimka',
        department: 'SocialNetwork',
    },
];

function getMemory(): Memory {
    return new Memory({
        data: flatData,
        keyProperty: 'key',
    });
}

function getPagingNavigation(
    hasMore: boolean = false,
    pageSize: number = 1,
    multiNavigation: boolean = false
): INavigationOptionValue<INavigationPageSourceConfig> {
    return {
        source: 'page',
        sourceConfig: {
            multiNavigation,
            pageSize,
            page: 0,
            hasMore,
        },
    };
}

const SliceContextInitValue = { slice: {} };
const SliceContext = createContext(SliceContextInitValue);

const SliceContextConnected = forwardRef(function (props, ref) {
    const [contextValue, setContextValue] = useState(SliceContextInitValue);
    const rerenderContext = useCallback(() => {
        setContextValue({ slice: clone(contextValue.slice) });
    }, []);

    useImperativeHandle(ref, () => rerenderContext);

    return <SliceContext.Provider value={contextValue}>{props.children}</SliceContext.Provider>;
});

function useSliceConnected<T extends keyof IListState>(props: {
    deps: T[];
    Slice;
    replacer?: (key: T, sliceState?: IListState) => unknown;
}) {
    const { Slice } = props;
    const forceRerenderRef = createRef();

    return {
        Slice: class extends Slice {
            constructor(config) {
                super({
                    ...config,
                    onChange: () => {
                        forceRerenderRef.current?.();
                    },
                });
                SliceContextInitValue.slice = this;
            }
            protected async __beforeApplyStateInternal(draftStateProp: T): Promise<T> {
                let actions;
                if (draftStateProp._actionToDispatch) {
                    actions = Array.from(draftStateProp._actionToDispatch.keys());
                }
                const newState = await super.__beforeApplyStateInternal(draftStateProp);
                newState.executedActionsToDispatch = actions;
                return newState;
            }
        },
        Component: () => (
            <SliceContextConnected ref={forceRerenderRef}>
                <TestComponent
                    deps={[...props.deps, 'executedActionsToDispatch']}
                    replacer={props.replacer}
                />
            </SliceContextConnected>
        ),
    };
}

function TestComponent<T extends keyof IListState>(props: {
    deps: T[];
    replacer?: (key: T, sliceState?: IListState) => unknown;
}) {
    const previousState = useRef<undefined | IListState>(undefined);
    const watchingPropertiesNames = useMemo(() => props.deps, []);
    const sliceContext = useContext(SliceContext) as unknown as { slice: ListSlice<IListState> };
    const [syncs, setSyncs] = useState<Record<T, unknown>[]>([]);

    useLayoutEffect(() => {
        const { slice } = sliceContext;
        setSyncs((prevSyncs) => {
            const resolveChanges = (keys, initValue) =>
                keys.reduce((acc, propertyName) => {
                    const newValue =
                        props?.replacer?.(propertyName, slice.state) ?? slice.state[propertyName];
                    const prevValue =
                        props?.replacer?.(propertyName, previousState?.current) ??
                        previousState?.current?.[propertyName];
                    // сравнивается содержание и упускается смена ссылки, что приводит к "пустым" рендерам
                    // поправить по задаче: https://online.sbis.ru/opendoc.html?guid=35e9b60d-a8bc-44b8-9b0f-b9907b957d93&client=3
                    if (!isEqual(newValue, prevValue) || !previousState?.current) {
                        acc[propertyName] = clone(newValue);
                    }
                    return acc;
                }, initValue as Record<T, unknown>);

            const watching = resolveChanges(watchingPropertiesNames, {});

            if (Object.keys(watching).length === 0) {
                const otherProps = resolveChanges(Object.keys(slice.state), {});
                const otherPropsKeys = Object.keys(otherProps);
                if (otherPropsKeys.length) {
                    watching.otherProps = otherPropsKeys;
                }
            }

            previousState.current = clone(slice.state);
            return [...prevSyncs, watching];
        });
    }, [sliceContext]);

    return (
        <div>
            {JSON.stringify(
                syncs,
                (key, value) => {
                    switch (typeof value) {
                        case 'function':
                            return 'function';
                        case 'undefined':
                            return 'undefined';
                        default:
                            return value;
                    }
                },
                4
            )}
        </div>
    );
}

describe('Controls/ListSliceSnapshots', () => {
    let container: HTMLDivElement;

    beforeEach(() => {
        container = document.createElement('div');
        WasabyEvents.initInstance(container);
        document.body.appendChild(container);
    });

    afterEach(() => {
        unmountComponentAtNode(container);
        WasabyEvents.destroyInstance(container);
        container.remove();
        container = null;
    });

    it('Во время метода load можно менять стейт, который не влияет на загрузку данных', async () => {
        const source = getMemory();
        const items = new RecordSet({
            rawData: flatData.slice(0, 2),
            keyProperty: 'key',
        });
        const sliceConfig = {
            source,
            filter: {
                department: 'Platform',
            },
        };
        const sliceLoadResult = {
            items,
        };
        const sliceConstructorParams = {
            config: sliceConfig,
            loadResult: sliceLoadResult,
        };

        const { Slice, Component } = useSliceConnected({
            deps: ['loading', 'markedKey'],
            Slice: ListSlice,
        });
        const slice = new Slice(sliceConstructorParams);

        const { asFragment } = render(<Component />, {
            container,
        });
        await waitFor(() => expect(slice.state).toBeDefined());
        act(() => {
            slice.load();
            slice.setState({ markedKey: 'test' });
        });

        await waitFor(() => expect(slice.state.loading).toBeTruthy());
        await waitFor(() => expect(slice.state.loading).toBeFalsy());
        expect(asFragment()).toMatchSnapshot();
    });

    it('При изменении фильтра должна произойти перезагрузка с новым фильтром', async () => {
        const items = new RecordSet({
            rawData: flatData,
            keyProperty: 'key',
        });
        const sliceConfig = {
            source: getMemory(),
            keyProperty: 'key',
            filter: {},
        };
        const sliceLoadResult = {
            items,
        };
        const sliceConstructorParams = {
            config: sliceConfig,
            loadResult: sliceLoadResult,
        };

        const { Slice, Component } = useSliceConnected({
            deps: ['loading', 'filter'],
            Slice: ListSlice,
        });
        const slice = new Slice(sliceConstructorParams);

        const { asFragment } = render(<Component />, {
            container,
        });

        await act(async () => {
            slice.setFilter({
                title: 'Платформа',
            });
        });

        await new Promise((resolve) =>
            setTimeout(() => {
                expect(asFragment()).toMatchSnapshot();
                resolve();
            }, 1000)
        );
    });

    it('После поиска не должно сбрасываться состояние searchInputValue, если оно менялось, пока выполнялся запрос', async () => {
        const items = new RecordSet();
        const sliceConfig = {
            source: getMemory(),
            searchParam: 'testSearchParam',
        };
        const sliceLoadResult = {
            items,
        };
        const sliceConstructorParams = {
            config: sliceConfig,
            loadResult: sliceLoadResult,
        };

        const { Slice, Component } = useSliceConnected({
            deps: ['loading', 'searchInputValue', 'searchValue', 'filter'],
            Slice: ListSlice,
        });
        const slice = new Slice(sliceConstructorParams);

        const { asFragment } = render(<Component />, {
            container,
        });
        await act(async () => {
            slice.search('test');
            await waitFor(() => expect(slice.state.loading).toBeTruthy());
            slice.setSearchInputValue('testPlus');
            await waitFor(() => expect(slice.state.loading).toBeFalsy());
        });

        expect(asFragment()).toMatchSnapshot();
    });

    it('Готово. Вызов reload с параметром keepNavigation должен сохранять навигацию при перезагрузке', async () => {
        const source = getMemory();
        const items = new RecordSet({
            rawData: [...flatData.slice(0, 1)],
            keyProperty: 'key',
        });
        const { Slice, Component } = useSliceConnected({
            deps: ['slice.state.items.getCount', 'loading', 'navigation'],
            replacer: (key, sliceState) => {
                if (key === 'slice.state.items.getCount') {
                    return sliceState?.items.getCount();
                }
                return undefined;
            },
            Slice: ListSlice,
        });
        const slice = new Slice({
            config: {
                source,
                navigation: getPagingNavigation(false, 1, false),
            },
            loadResult: {
                items,
            },
        });

        const { asFragment } = render(<Component />, {
            container,
        });

        act(() => {
            slice.load('down');
        });
        await waitFor(() => expect(slice.state.loading).toBeTruthy());
        await waitFor(() => expect(slice.state.loading).toBeFalsy());

        act(() => {
            slice.reload(null, true);
        });
        await waitFor(() => expect(slice.state.loading).toBeTruthy());
        await waitFor(() => expect(slice.state.loading).toBeFalsy());

        expect(asFragment()).toMatchSnapshot();
    });

    it('Вызов reload в момент, когда уже идёт загрузка. Стейт loading из-за отмены предыдущей загрузки не должен сбрасываться ', async () => {
        const source = getMemory();
        const items = new RecordSet({
            keyProperty: 'id',
            rawData: flatData,
        });

        const { Slice, Component } = useSliceConnected({
            deps: ['loading', 'filter'],
            Slice: ListSlice,
        });

        const slice: ListSlice = new Slice({
            config: {
                source,
            },
            loadResult: {
                items,
            },
        });

        const { asFragment } = render(<Component />, {
            container,
        });

        act(() => {
            slice.setState({ filter: { key: 1 } });
            slice.reload();
        });

        await waitFor(() => expect(slice.state.loading).toBeTruthy());
        await waitFor(() => expect(slice.state.loading).toBeFalsy());

        expect(asFragment()).toMatchSnapshot();
    });

    it('selectionViewMode сбрасывается при перезагрузке (вызов метода reload)', async () => {
        const source = getMemory();
        const items = new RecordSet({
            keyProperty: 'id',
            rawData: flatData,
        });

        const { Slice, Component } = useSliceConnected({
            deps: ['loading', 'selectionViewMode'],
            Slice: ListSlice,
        });

        const slice = new Slice({
            config: {
                source,
            },
            loadResult: {
                items,
            },
        });

        const { asFragment } = render(<Component />, {
            container,
        });

        act(() => {
            slice.setSelectionViewMode('selected');
        });
        act(() => {
            slice.reload();
        });

        await waitFor(() => expect(slice.state.loading).toBeTruthy());
        await waitFor(() => expect(slice.state.loading).toBeFalsy());

        expect(asFragment()).toMatchSnapshot();
    });

    it('повторный вызов метода reload должен отменить предыдущую загрузку, если она выполняется', async () => {
        const source = getMemory();
        const items = new RecordSet({
            keyProperty: 'id',
            rawData: flatData,
        });

        const { Slice, Component } = useSliceConnected({
            deps: ['loading', 'selectionViewMode'],
            Slice: ListSlice,
        });

        const slice = new Slice({
            config: {
                source,
            },
            loadResult: {
                items,
            },
        });

        const { asFragment } = render(<Component />, {
            container,
        });

        await act(async () => {
            const promise = slice.reload();
            await waitFor(() => expect(slice.state.loading).toBeTruthy());
            slice.reload();
            promise.catch((error) => {
                expect(error.isCanceled).toBeTruthy();
            });
        });

        await waitFor(() => expect(slice.state.loading).toBeFalsy());

        expect(asFragment()).toMatchSnapshot();
    });

    it('reload должен выполнить перезагрузку и обновить items на стейте', async () => {
        const source = getMemory();
        const items = new RecordSet();

        const { Slice, Component } = useSliceConnected({
            deps: ['items', 'loading'],
            replacer: (key, sliceState) => {
                if (key === 'items') {
                    return sliceState?.items.getCount();
                }
            },
            Slice: ListSlice,
        });

        const slice = new Slice({
            config: {
                source,
            },
            loadResult: {
                items,
            },
        });

        const { asFragment } = render(<Component />, {
            container,
        });

        act(() => {
            slice.reload();
        });

        await waitFor(() => expect(slice.state.loading).toBeTruthy());
        await waitFor(() => expect(slice.state.loading).toBeFalsy());

        expect(asFragment()).toMatchSnapshot();
    });

    it('Вызов метода search c тем же searchValue, что и на стейте', async () => {
        const source = new Memory();
        const items = new RecordSet();

        const { Slice, Component } = useSliceConnected({
            deps: ['loading', 'searchValue'],
            Slice: ListSlice,
        });

        const slice = new Slice({
            config: {
                source,
                dataLoadedReturnPromise: false,
                searchParam: 'testSearchParam',
                searchValue: 'test',
            },
            loadResult: {
                items,
            },
        });

        const { asFragment } = render(<Component />, {
            container,
        });

        await act(async () => {
            slice.search('test');
            await waitFor(() => expect(slice.state.loading).toBeTruthy());
            slice.search('test');
            await waitFor(() => expect(slice.state.loading).toBeFalsy());
        });

        expect(asFragment()).toMatchSnapshot();
    });

    describe('Загрузка записей по скролу', () => {
        it('Источник вернул ошибку загрузки', async () => {
            const error = new fetch.Errors.HTTP({
                httpError: HTTPStatus.GatewayTimeout,
                message: 'test',
                url: 'test',
            });
            const source = new Memory();
            const items = new RecordSet();

            source.query = () => Promise.reject(error);

            const { Slice, Component } = useSliceConnected({
                deps: ['errorViewConfig'],
                Slice: ListSlice,
                replacer: (key, sliceState) => {
                    if (key === 'errorViewConfig' && sliceState?.errorViewConfig) {
                        return {
                            ...sliceState.errorViewConfig,
                            options: {
                                ...sliceState.errorViewConfig.options,
                                error: null,
                            },
                        };
                    }
                },
            });

            const slice = new Slice({
                config: {
                    source,
                },
                loadResult: {
                    items,
                },
            });

            const { asFragment } = render(<Component />, {
                container,
            });

            await act(async () => {
                return slice._loadItemsToDirection('down').catch((error) => error);
            });

            await waitFor(() => expect(slice.state.errorViewConfig.mode).toStrictEqual('inlist'));

            expect(asFragment()).toMatchSnapshot();
        });
    });
});
