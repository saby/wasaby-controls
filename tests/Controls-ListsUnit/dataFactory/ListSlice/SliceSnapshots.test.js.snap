// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls/ListSliceSnapshots reload должен выполнить перезагрузку и обновить items на стейте 1`] = `
<DocumentFragment>
  <div>
    [
    {
        "items": 0,
        "loading": false,
        "executedActionsToDispatch": "undefined"
    },
    {
        "loading": true
    },
    {
        "executedActionsToDispatch": [
            "reload"
        ]
    },
    {
        "items": 3,
        "loading": false,
        "executedActionsToDispatch": [
            "newItemsReceived"
        ]
    }
]
  </div>
</DocumentFragment>
`;

exports[`Controls/ListSliceSnapshots selectionViewMode сбрасывается при перезагрузке (вызов метода reload) 1`] = `
<DocumentFragment>
  <div>
    [
    {
        "loading": false,
        "selectionViewMode": "hidden",
        "executedActionsToDispatch": "undefined"
    },
    {
        "selectionViewMode": "selected"
    },
    {
        "loading": true
    },
    {
        "executedActionsToDispatch": [
            "reload"
        ]
    },
    {
        "loading": false,
        "executedActionsToDispatch": [
            "newItemsReceived"
        ]
    }
]
  </div>
</DocumentFragment>
`;

exports[`Controls/ListSliceSnapshots Во время метода load можно менять стейт, который не влияет на загрузку данных 1`] = `
<DocumentFragment>
  <div>
    [
    {
        "loading": false,
        "markedKey": "undefined",
        "executedActionsToDispatch": "undefined"
    },
    {
        "loading": true
    },
    {
        "executedActionsToDispatch": [
            "oldSliceLoad"
        ]
    },
    {
        "loading": false,
        "executedActionsToDispatch": "undefined"
    }
]
  </div>
</DocumentFragment>
`;

exports[`Controls/ListSliceSnapshots Вызов reload в момент, когда уже идёт загрузка. Стейт loading из-за отмены предыдущей загрузки не должен сбрасываться  1`] = `
<DocumentFragment>
  <div>
    [
    {
        "loading": false,
        "filter": {},
        "executedActionsToDispatch": "undefined"
    },
    {
        "loading": true
    },
    {},
    {
        "loading": false,
        "filter": {
            "key": 1
        }
    },
    {
        "loading": true
    },
    {
        "executedActionsToDispatch": [
            "reload"
        ]
    },
    {
        "loading": false,
        "executedActionsToDispatch": [
            "newItemsReceived"
        ]
    }
]
  </div>
</DocumentFragment>
`;

exports[`Controls/ListSliceSnapshots Вызов reload с параметром keepNavigation должен сохранять навигацию при перезагрузке 1`] = `
<DocumentFragment>
  <div>
    [
    {
        "slice.state.items.getCount": 1,
        "loading": false,
        "navigation": {
            "source": "page",
            "sourceConfig": {
                "multiNavigation": false,
                "pageSize": 1,
                "page": 0,
                "hasMore": false
            }
        },
        "executedActionsToDispatch": "undefined"
    },
    {
        "loading": true
    },
    {
        "executedActionsToDispatch": [
            "oldSliceLoad"
        ]
    },
    {
        "loading": false,
        "executedActionsToDispatch": "undefined"
    },
    {
        "loading": true
    },
    {
        "executedActionsToDispatch": [
            "reload"
        ]
    },
    {
        "loading": false,
        "executedActionsToDispatch": [
            "newItemsReceived"
        ]
    }
]
  </div>
</DocumentFragment>
`;

exports[`Controls/ListSliceSnapshots Вызов метода load без аргументов. Фильтр не должен меняться. 1`] = `
<DocumentFragment>
  <div>
    [
    {
        "items": 2,
        "loading": false,
        "filter": {
            "department": "Platform"
        },
        "executedActionsToDispatch": "undefined"
    },
    {
        "loading": true
    },
    {
        "executedActionsToDispatch": [
            "oldSliceLoad"
        ]
    },
    {
        "loading": false,
        "executedActionsToDispatch": "undefined"
    }
]
  </div>
</DocumentFragment>
`;

exports[`Controls/ListSliceSnapshots Вызов метода search c тем же searchValue, что и на стейте 1`] = `
<DocumentFragment>
  <div>
    [
    {
        "loading": false,
        "searchValue": "test",
        "executedActionsToDispatch": "undefined"
    },
    {
        "loading": true
    },
    {
        "executedActionsToDispatch": [
            "reload"
        ]
    },
    {
        "loading": false,
        "executedActionsToDispatch": [
            "newItemsReceived"
        ]
    }
]
  </div>
</DocumentFragment>
`;

exports[`Controls/ListSliceSnapshots Повторный вызов метода reload должен отменить предыдущую загрузку, если она выполняется 1`] = `
<DocumentFragment>
  <div>
    [
    {
        "loading": false,
        "selectionViewMode": "hidden",
        "executedActionsToDispatch": "undefined"
    },
    {
        "loading": true
    },
    {
        "executedActionsToDispatch": [
            "reload"
        ]
    },
    {
        "loading": false,
        "executedActionsToDispatch": [
            "newItemsReceived"
        ]
    }
]
  </div>
</DocumentFragment>
`;

exports[`Controls/ListSliceSnapshots После поиска не должно сбрасываться состояние searchInputValue, если оно менялось, пока выполнялся запрос 1`] = `
<DocumentFragment>
  <div>
    [
    {
        "loading": false,
        "searchInputValue": "",
        "searchValue": "",
        "filter": {},
        "executedActionsToDispatch": "undefined"
    },
    {
        "searchInputValue": "test"
    },
    {},
    {
        "loading": true
    },
    {
        "searchInputValue": "testPlus"
    },
    {
        "loading": false,
        "searchValue": "test",
        "filter": {
            "testSearchParam": "test"
        }
    }
]
  </div>
</DocumentFragment>
`;

exports[`Controls/ListSliceSnapshots При изменении фильтра должна произойти перезагрузка с новым фильтром 1`] = `
<DocumentFragment>
  <div>
    [
    {
        "loading": false,
        "filter": {},
        "executedActionsToDispatch": "undefined"
    },
    {
        "loading": true
    },
    {},
    {
        "loading": false,
        "filter": {
            "title": "Платформа"
        }
    }
]
  </div>
</DocumentFragment>
`;
